### DISCUSSION TITLE
New SRTP Aliases constraints giving "Invalid constraint."

### DISCUSSION THREAD

**Discussion by TheAngryByrd** [dotnet/fsharp#15567] [Q&A]
👋  [F# 7 introduced new ways of working with SRTP aliases](https://devblogs.microsoft.com/dotnet/announcing-fsharp-7/#making-working-with-srtps-easier) however I've ran into an issue converting some type contraints.

Below is some code from [IcedTask](https://github.com/TheAngryByrd/IcedTasks) that uses this feature to create the Awaiter/Awaitable structures with SRTP aliases. 

In the `FooBuilder` I show what I think are two implementations that should result in the same behavior but they're not.

---

```fsharp

open System.Runtime.CompilerServices

/// A structure that looks like an Awaiter
type Awaiter<'Awaiter, 'TResult
    when 'Awaiter :> ICriticalNotifyCompletion
    and 'Awaiter: (member get_IsCompleted: unit -> bool)
    and 'Awaiter: (member GetResult: unit -> 'TResult)> = 'Awaiter

/// Functions for Awaiters
type Awaiter =
    /// Gets a value that indicates whether the asynchronous task has completed
    static member inline IsCompleted<'Awaiter, 'TResult when Awaiter<'Awaiter, 'TResult>>
        (awaiter: 'Awaiter)
        =
        awaiter.get_IsCompleted ()

    /// Ends the wait for the completion of the asynchronous task.
    static member inline GetResult<'Awaiter, 'TResult when Awaiter<'Awaiter, 'TResult>>
        (awaiter: 'Awaiter)
        =
        awaiter.GetResult()


    /// Schedules the continuation action that's invoked when the instance completes
    static member inline OnCompleted<'Awaiter, 'TResult, 'Continuation
        when Awaiter<'Awaiter, 'TResult>>
        (
            awaiter: 'Awaiter,
            continuation: System.Action
        ) =
        awaiter.OnCompleted(continuation)

    /// Schedules the continuation action that's invoked when the instance completes.
    static member inline UnsafeOnCompleted<'Awaiter, 'TResult, 'Continuation
        when Awaiter<'Awaiter, 'TResult>>
        (
            awaiter: 'Awaiter,
            continuation: System.Action
        ) =
        awaiter.UnsafeOnCompleted(continuation)

/// A structure looks like an Awaitable
type Awaitable<'Awaitable, 'Awaiter, 'TResult
    when 'Awaitable: (member GetAwaiter: unit -> Awaiter<'Awaiter, 'TResult>)> = 'Awaitable

/// Functions for Awaitables
type Awaitable =
    /// Creates an awaiter for this value.
    static member inline GetAwaiter<'Awaitable, 'Awaiter, 'TResult
        when Awaitable<'Awaitable, 'Awaiter, 'TResult>>
        (awaitable: 'Awaitable)
        =
        awaitable.GetAwaiter()

open Microsoft.FSharp.Core.CompilerServices
open System.Threading

type FooBuilder() =

    // Compiles
    [<NoEagerConstraintApplication>]
    member inline _.Source<'Awaitable, 'Awaiter, 'TResult1, 'Error
        when ^Awaitable: (member GetAwaiter: unit -> ^Awaiter)
        and ^Awaiter :> ICriticalNotifyCompletion
        and ^Awaiter: (member get_IsCompleted: unit -> bool)
        and ^Awaiter: (member GetResult: unit -> Result<'TResult1, 'Error>)>

        (task: 'Awaitable)
        : CancellationToken -> ^Awaiter =
        (fun (ct: CancellationToken) -> Awaitable.GetAwaiter task)


    // Doesn't
    // Invalid constraint. Valid constraint forms include "'T :> ISomeInterface" for interface constraints and "SomeConstrainingType<'T>" for self-constraints. See https://aka.ms/fsharp-type-constraints.F# Compiler3533
    [<NoEagerConstraintApplication>]
    member inline _.Source<'Awaitable, 'Awaiter, 'TResult1, 'Error
        when Awaitable<'Awaitable, 'Awaiter, Result<'TResult1, 'Error>>>

        (task: 'Awaitable)
        : CancellationToken -> ^Awaiter =
        (fun (ct: CancellationToken) -> Awaitable.GetAwaiter task)

```

--- 

What am I doing wrong here or is this maybe a bug or maybe a refinement needed?

**Comment by En3Tho** [comment]
Is it supposed to work with multiple generic parameters like this? Does `when 'Awaitble :> Awaitable<'Awaitable, ...>` work?

**Reply by TheAngryByrd** [reply]
> Is it supposed to work with multiple generic parameters like this?

I should clarify, it does without the `Result<'TResult1, 'Error>`. Such as `when Awaitable<'Awaitable, 'Awaiter, 'TResult>` works and `when Awaitable<'Awaitable, 'Awaiter, Result<'TResult1, 'Error>>>` doesn't work when constraining the `'Result` further.


> Does when 'Awaitble :> Awaitable<'Awaitable, ...> work?

Results in `invalid constraint: the type used for the constraint is sealed, which means the constraint could only be satisfied by at most one solution`

**Reply by En3Tho** [reply]
I see. So the presence of Result<..., ...> breaks it. It looks like it expects generics only?

As a workaround you can use something like this I guess (well, maybe you already know this). It produces a warning (missing constraints blablabla) but is still working
```fsharp
[<NoEagerConstraintApplication>]
    member inline _.Source2<'Awaitable, 'Awaiter, 'TResult1, 'Error>
        (task: Awaitable<'Awaitable, 'Awaiter, Result<'TResult1, 'Error>>)
        : CancellationToken -> ^Awaiter =
        (fun (ct: CancellationToken) -> Awaitable.GetAwaiter task)
```

**Reply by TheAngryByrd** [reply]
> I see. So the presence of Result<..., ...> breaks it. It looks like it expects generics only?

Yeah that seems to be the case. Not sure if this is intended or oversight in the new syntax or what.

> 
> As a workaround you can use something like this I guess (well, maybe you already know this). It produces a warning (missing constraints blablabla) but is still working
> 
> ```fsharp
> [<NoEagerConstraintApplication>]
>     member inline _.Source2<'Awaitable, 'Awaiter, 'TResult1, 'Error>
>         (task: Awaitable<'Awaitable, 'Awaiter, Result<'TResult1, 'Error>>)
>         : CancellationToken -> ^Awaiter =
>         (fun (ct: CancellationToken) -> Awaitable.GetAwaiter task)
> ```

I seem to get more errors with this:

```
A type parameter is missing a constraint 'when ^Awaitable: (member GetAwaiter: unit -> Awaiter<'?201918439,'?201918440>)'F# Compiler[193](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/compiler-messages/fs0193)
A type parameter is missing a constraint 'when ^Awaiter :> ICriticalNotifyCompletion'F# Compiler[193](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/compiler-messages/fs0193)
A type parameter is missing a constraint 'when ^Awaiter: (member get_IsCompleted: unit -> bool)'F# Compiler[193](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/compiler-messages/fs0193)
A type parameter is missing a constraint 'when ^Awaiter: (member GetResult: unit -> '?201918440)'F# Compiler[193](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/compiler-messages/fs0193)
'Awaitable (requires member GetAwaiter)
```

**Reply by vzarytovskii** [reply]
> Yeah that seems to be the case. Not sure if this is intended or oversight in the new syntax or what.

I would expect it's intentional.

Could you put an additional constraint on it instead like `and 'Result :> Result<... >`?

Sorry, typing from phone.

**Reply by vzarytovskii** [reply]
Like
```fsharp
[<NoEagerConstraintApplication>]
member inline _.Source<'Awaitable, 'Awaiter, 'TResult1, 'Error when Awaitable<'Awaitable, 'Awaiter ,'Result> and 'Result :> Result<'TResult1, 'Error>> (task: 'Awaitable) : CancellationToken -> ^Awaiter =
    (fun (ct: CancellationToken) -> Awaitable.GetAwaiter task)

**Reply by vzarytovskii** [reply]
On the second thought, that probably won't work.

**Reply by TheAngryByrd** [reply]
@vzarytovskii you're on vacation, stop it!

**Reply by En3Tho** [reply]
> I seem to get more errors with this:

I guess you have `Treat warnings as errors` option on? It works on my machine, just produces warnings you've described (and I'm not sure what is the point of producing those, really)

**Reply by TheAngryByrd** [reply]
Yeah you're right I have WarnAsError on, however taking that off doesn't solve the issue as eventually I get this:

```
This construct causes code to be less generic than indicated by the type annotations. The type variable 'TaskLike has been constrained to be type 'obj'.F# Compiler[64](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/compiler-messages/fs0064)
```
